### ISSUE TITLE
Pipeline Execution Does Not Work With us-east-2 & Cognito

### ISSUE THREAD

**Issue by AlexandreBrown** [awslabs/kubeflow-manifests#148]
**Describe the bug**
After running a kubeflow pipeline with S3 as `pipeline_root`, the pipeline run fails.

**Steps To Reproduce**
1. Create a cluster in **us-east-2** using EKS 1.21
2. Create a bucket in **us-east-2**
3. Install Kubeflow using `main` + `v1.4.1` upstream following `cognito-rds-s3` + KFServing (not sure if KFServing changes anything, probably not) setup.
4. Setup secrets access for your namespace following the [official guide](https://github.com/awslabs/kubeflow-manifests/blob/main/docs/component-guides/notebooks.md#setup-rds--s3-credentials-access-for-notebooks--pipelines)
5. Create an empty KFP Experiment using Kubeflow UI
6. Create and run the pipeline under `kfp 1.8.X`
```python
from kfp.v2.dsl import component
from kfp.v2.dsl import pipeline
from kfp import Client
from kfp.dsl import PipelineExecutionMode
from kfp.dsl import PipelineConf
from kfp.aws import use_aws_secret
from datetime import datetime


@component()
def my_component():
    print("hello")


@pipeline(
    name="my-pipeline",
    description="my_pipeline",
    pipeline_root="s3://YOUR_BUCKET/experiments"
)
def my_pipeline():
    my_component().set_caching_options(False)


kubeflow_gateway_endpoint = "YOUR_ENDPOINT"

alb_session_cookie0="YOUR_VALUE_0"
alb_session_cookie1="YOUR_VALUE_1"
client = Client(
    host=f"https://{kubeflow_gateway_endpoint}/pipeline", 
    cookies=f"AWSELBAuthSessionCookie-0={alb_session_cookie0};AWSELBAuthSessionCookie-1={alb_session_cookie1}"
)

NAMESPACE = "YOUR_NAMESPACE"
EXPERIMENT_NAME = "YOUR_EXPERIMENT"
run_name = f"pipeline-test-{datetime.now()}"


pipeline_conf = PipelineConf()
pipeline_conf.add_op_transformer(use_aws_secret(
    secret_name='mlpipeline-minio-artifact',
    aws_access_key_id_name='accesskey',
    aws_secret_access_key_name='secretkey',
    aws_region='us-east-2'
))
client.create_run_from_pipeline_func(
    my_pipeline,
    arguments={},
    run_name=run_name,
    experiment_name=EXPERIMENT_NAME,
    pipeline_conf=pipeline_conf,
    namespace=NAMESPACE,
    mode=PipelineExecutionMode.V2_COMPATIBLE
)
```
**Expected behavior**
The pipeline should succeed.

**Environment**
* Kubernetes version EKS 1.21
* Using EKS (yes/no), if so version? Yes
* Kubeflow version 1.4.1
* AWS build number: AWS main
* AWS service targeted (S3, RDS, etc.): NA

**Screenshots**
![KfpError](https://user-images.githubusercontent.com/26939775/159100346-48131343-9ffc-4b4f-a231-32dbdd372c4a.png)
```
This step is in Error state with this message: Error (exit code 1): failed to put file: Put "http://YOUR_BUCKET.s3.dualstack.us-east-1.amazonaws.com/artifacts/my-pipeline-khs2z/my-pipeline-khs2z-2408580201/main.log": 301 response missing Location header
```

**Additional info**  
- If we create a component that has some `Output` artifacts (eg: `Output[Dataset]`) then then artifact will be pushed to S3 successfully but still not the `main.log` making the pipeline fail.  
- Pipeline fails even with KFP sdk v1 for the same reason.  
- Same failure happens if I compile the pipeline and upload it manually via KF UI  

`kubectl describe pod my-pipeline -n dev`  
  
```
Name:         my-pipeline-khs2z-2408580201
Namespace:    dev
Priority:     0
Node:         ip-xxx-xxx-xx-xxx.us-east-2.compute.internal/xxx.xxx.xx.xxx
Start Time:   Sat, 19 Mar 2022 00:46:32 +0000
Labels:       pipeline/runid=62c315bb-9cd5-44a3-abbc-f98d5dff08e1
              pipelines.kubeflow.org/cache_enabled=true
              pipelines.kubeflow.org/enable_caching=false
              pipelines.kubeflow.org/kfp_sdk_version=1.8.6
              pipelines.kubeflow.org/pipeline-sdk-type=kfp
              pipelines.kubeflow.org/v2_component=true
              workflows.argoproj.io/completed=true
              workflows.argoproj.io/workflow=my-pipeline-khs2z
Annotations:  kubernetes.io/psp: eks.privileged
              pipelines.kubeflow.org/component_ref: {}
              pipelines.kubeflow.org/v2_component: true
              sidecar.istio.io/inject: false
              workflows.argoproj.io/node-name: my-pipeline-khs2z.my-component
              workflows.argoproj.io/template:
                {"name":"my-component","inputs":{"parameters":[{"name":"pipeline-name","value":"pipeline/my-pipeline"},{"name":"pipeline-root","value":"s3...
Status:       Failed
IP:           xxx.xxx.xx.xxx
IPs:
  IP:           xxx.xxx.xx.xxx
Controlled By:  Workflow/my-pipeline-khs2z
Init Containers:
  kfp-launcher:
    Container ID:  docker://6453024929c9d97410499d66f851b01bbea239054d57692e54a7b89315db34cf
    Image:         gcr.io/ml-pipeline/kfp-launcher:1.8.6
    Image ID:      docker-pullable://gcr.io/ml-pipeline/kfp-launcher@sha256:608f554a8b63a917199856e40253721c6f1af831f0fa9769a37be6945b3af936
    Port:          <none>
    Host Port:     <none>
    Command:
      launcher
      --copy
      /kfp-launcher/launch
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Sat, 19 Mar 2022 00:46:35 +0000
      Finished:     Sat, 19 Mar 2022 00:46:35 +0000
    Ready:          True
    Restart Count:  0
    Environment:
      ARGO_CONTAINER_NAME:         kfp-launcher
      ARGO_INCLUDE_SCRIPT_OUTPUT:  false
    Mounts:
      /kfp-launcher from kfp-launcher (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4hb8l (ro)
Containers:
  wait:
    Container ID:  docker://f5b6073cd2915715f26f96e910926d3f000f39e6[6829](https://gitlab.com/geown/artificial-intelligence/applications/kubeflow/-/jobs/2224774677#L6829)ead5881b25b22473b873
    Image:         gcr.io/ml-pipeline/argoexec:v3.1.6-patch-license-compliance
    Image ID:      docker-pullable://gcr.io/ml-pipeline/argoexec@sha256:44cf8455a51aa5b961d1a86f65e39adf5ffca9bdcd33a745c3b79f430b7439e0
    Port:          <none>
    Host Port:     <none>
    Command:
      argoexec
      wait
      --loglevel
      info
    State:          Terminated
      Reason:       Error
      Message:      failed to put file: Put "http://YOUR_BUCKET.s3.dualstack.us-east-1.amazonaws.com/artifacts/my-pipeline-khs2z/my-pipeline-khs2z-2408580201/main.log": 301 response missing Location header
      Exit Code:    1
      Started:      Sat, 19 Mar 2022 00:46:43 +0000
      Finished:     Sat, 19 Mar 2022 00:48:18 +0000
    Ready:          False
    Restart Count:  0
    Environment:
      ARGO_POD_NAME:                    my-pipeline-khs2z-2408580201 (v1:metadata.name)
      ARGO_CONTAINER_RUNTIME_EXECUTOR:  
      GODEBUG:                          x509ignoreCN=0
      ARGO_CONTAINER_NAME:              wait
      ARGO_INCLUDE_SCRIPT_OUTPUT:       false
    Mounts:
      /argo/podmetadata from podmetadata (rw)
      /argo/secret/mlpipeline-minio-artifact from mlpipeline-minio-artifact (ro)
      /mainctrfs/kfp-launcher from kfp-launcher (rw)
      /var/run/docker.sock from docker-sock (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4hb8l (ro)
  main:
    Container ID:  docker://c3fe466dfa3b5dc0b738a01c00e26a02789eff4def70d1dbafe697d3486cfec1
    Image:         python:3.7
    Image ID:      docker-pullable://python@sha256:64b89c9361673c4e298604901b58b407404134e7f6c70e871fd53695baa96ca1
    Port:          <none>
    Host Port:     <none>
    Command:
      /kfp-launcher/launch
      --mlmd_server_address
      $(METADATA_GRPC_SERVICE_HOST)
      --mlmd_server_port
      $(METADATA_GRPC_SERVICE_PORT)
      --runtime_info_json
      $(KFP_V2_RUNTIME_INFO)
      --container_image
      $(KFP_V2_IMAGE)
      --task_name
      my-component
      --pipeline_name
      pipeline/my-pipeline
      --run_id
      $(KFP_RUN_ID)
      --run_resource
      workflows.argoproj.io/$(WORKFLOW_ID)
      --namespace
      $(KFP_NAMESPACE)
      --pod_name
      $(KFP_POD_NAME)
      --pod_uid
      $(KFP_POD_UID)
      --pipeline_root
      s3://YOUR_BUCKET/experiments/image/segmentation/baseline
      --enable_caching
      $(ENABLE_CACHING)
      --
      --
    Args:
      sh
      -c
      
      if ! [ -x "$(command -v pip)" ]; then
          python3 -m ensurepip || python3 -m ensurepip --user || apt-get install python3-pip
      fi
      
      PIP_DISABLE_PIP_VERSION_CHECK=1 python3 -m pip install --quiet     --no-warn-script-location 'kfp==1.8.6' && "$0" "$@"
      
      sh
      -ec
      program_path=$(mktemp -d)
      printf "%s" "$0" > "$program_path/ephemeral_component.py"
      python3 -m kfp.v2.components.executor_main                         --component_module_path                         "$program_path/ephemeral_component.py"                         "$@"
      
      
      import kfp
      from kfp.v2 import dsl
      from kfp.v2.dsl import *
      from typing import *
      
      def my_component():
          print("hello")
      
      
      --executor_input
      {{$}}
      --function_to_execute
      my_component
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Sat, 19 Mar 2022 00:46:44 +0000
      Finished:     Sat, 19 Mar 2022 00:47:01 +0000
    Ready:          False
    Restart Count:  0
    Environment Variables from:
      metadata-grpc-configmap  ConfigMap  Optional: true
    Environment:
      AWS_ACCESS_KEY_ID:           <set to the key 'accesskey' in secret 'mlpipeline-minio-artifact'>  Optional: false
      AWS_SECRET_ACCESS_KEY:       <set to the key 'secretkey' in secret 'mlpipeline-minio-artifact'>  Optional: false
      AWS_REGION:                  us-east-2
      KFP_POD_NAME:                my-pipeline-khs2z-2408580201 (v1:metadata.name)
      KFP_POD_UID:                  (v1:metadata.uid)
      KFP_NAMESPACE:               dev (v1:metadata.namespace)
      WORKFLOW_ID:                  (v1:metadata.labels['workflows.argoproj.io/workflow'])
      KFP_RUN_ID:                   (v1:metadata.labels['pipeline/runid'])
      ENABLE_CACHING:               (v1:metadata.labels['pipelines.kubeflow.org/enable_caching'])
      KFP_V2_IMAGE:                python:3.7
      KFP_V2_RUNTIME_INFO:         {"inputParameters": {}, "inputArtifacts": {}, "outputParameters": {}, "outputArtifacts": {}}
      ARGO_CONTAINER_NAME:         main
      ARGO_INCLUDE_SCRIPT_OUTPUT:  false
    Mounts:
      /kfp-launcher from kfp-launcher (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4hb8l (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  podmetadata:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.annotations -> annotations
  docker-sock:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/docker.sock
    HostPathType:  Socket
  kfp-launcher:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  mlpipeline-minio-artifact:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  mlpipeline-minio-artifact
    Optional:    false
  kube-api-access-4hb8l:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>
```  
  
`kubectl get ingress -n istio-system istio-ingress`
```
NAME            CLASS    HOSTS   ADDRESS                                                                 PORTS   AGE
istio-ingress   <none>   *       xxx.us-east-2.elb.amazonaws.com   80      2d13h
```

**Comment by AlexandreBrown** [comment]
@surajkota Maybe it has to do with the fact I have created a new profile? Not sure if `apps/pipeline/params.env` will cover all profiles?

**Comment by AlexandreBrown** [comment]
`kubectl describe pod my-pipeline -n my-namespace`  
  
```
Name:         my-pipeline-khs2z-2408580201
Namespace:    my-namespace
Priority:     0
Node:         ip-xxx-xxx-xx-xxx.us-east-2.compute.internal/xxx.1xxx.xx.xxx
Start Time:   Sat, 19 Mar 2022 00:46:32 +0000
Labels:       pipeline/runid=62c315bb-9cd5-44a3-abbc-f98d5dff08e1
              pipelines.kubeflow.org/cache_enabled=true
              pipelines.kubeflow.org/enable_caching=false
              pipelines.kubeflow.org/kfp_sdk_version=1.8.6
              pipelines.kubeflow.org/pipeline-sdk-type=kfp
              pipelines.kubeflow.org/v2_component=true
              workflows.argoproj.io/completed=true
              workflows.argoproj.io/workflow=my-pipeline-khs2z
Annotations:  kubernetes.io/psp: eks.privileged
              pipelines.kubeflow.org/component_ref: {}
              pipelines.kubeflow.org/v2_component: true
              sidecar.istio.io/inject: false
              workflows.argoproj.io/node-name: my-pipeline-khs2z.my-component
              workflows.argoproj.io/template:
                {"name":"my-component","inputs":{"parameters":[{"name":"pipeline-name","value":"pipeline/my-pipeline"},{"name":"pipeline-root","value":"s3...
Status:       Failed
IP:           xxx.xxx.xx.xxx
IPs:
  IP:           xxx.xxx.xx.xxx
Controlled By:  Workflow/my-pipeline-khs2z
Init Containers:
  kfp-launcher:
    Container ID:  docker://6453024929c9d97410499d66f851b01bbea239054d57692e54a7b89315db34cf
    Image:         gcr.io/ml-pipeline/kfp-launcher:1.8.6
    Image ID:      docker-pullable://gcr.io/ml-pipeline/kfp-launcher@sha256:608f554a8b63a917199856e40253721c6f1af831f0fa9769a37be6945b3af936
    Port:          <none>
    Host Port:     <none>
    Command:
      launcher
      --copy
      /kfp-launcher/launch
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Sat, 19 Mar 2022 00:46:35 +0000
      Finished:     Sat, 19 Mar 2022 00:46:35 +0000
    Ready:          True
    Restart Count:  0
    Environment:
      ARGO_CONTAINER_NAME:         kfp-launcher
      ARGO_INCLUDE_SCRIPT_OUTPUT:  false
    Mounts:
      /kfp-launcher from kfp-launcher (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4hb8l (ro)
Containers:
  wait:
    Container ID:  docker://f5b6073cd2915715f26f96e910926d3f000f39e6[6829](https://gitlab.com/geown/artificial-intelligence/applications/kubeflow/-/jobs/2224774677#L6829)ead5881b25b22473b873
    Image:         gcr.io/ml-pipeline/argoexec:v3.1.6-patch-license-compliance
    Image ID:      docker-pullable://gcr.io/ml-pipeline/argoexec@sha256:44cf8455a51aa5b961d1a86f65e39adf5ffca9bdcd33a745c3b79f430b7439e0
    Port:          <none>
    Host Port:     <none>
    Command:
      argoexec
      wait
      --loglevel
      info
    State:          Terminated
      Reason:       Error
      **Message:      failed to put file: Put "http://YOUR_BUCKET.s3.dualstack.us-east-1.amazonaws.com/artifacts/my-pipeline-khs2z/my-pipeline-khs2z-2408580201/main.log": 301 response missing Location header**
      Exit Code:    1
      Started:      Sat, 19 Mar 2022 00:46:43 +0000
      Finished:     Sat, 19 Mar 2022 00:48:18 +0000
    Ready:          False
    Restart Count:  0
    Environment:
      ARGO_POD_NAME:                    my-pipeline-khs2z-2408580201 (v1:metadata.name)
      ARGO_CONTAINER_RUNTIME_EXECUTOR:  
      GODEBUG:                          x509ignoreCN=0
      ARGO_CONTAINER_NAME:              wait
      ARGO_INCLUDE_SCRIPT_OUTPUT:       false
    Mounts:
      /argo/podmetadata from podmetadata (rw)
      /argo/secret/mlpipeline-minio-artifact from mlpipeline-minio-artifact (ro)
      /mainctrfs/kfp-launcher from kfp-launcher (rw)
      /var/run/docker.sock from docker-sock (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4hb8l (ro)
  main:
    Container ID:  docker://c3fe466dfa3b5dc0b738a01c00e26a02789eff4def70d1dbafe697d3486cfec1
    Image:         python:3.7
    Image ID:      docker-pullable://python@sha256:64b89c9361673c4e298604901b58b407404134e7f6c70e871fd53695baa96ca1
    Port:          <none>
    Host Port:     <none>
    Command:
      /kfp-launcher/launch
      --mlmd_server_address
      $(METADATA_GRPC_SERVICE_HOST)
      --mlmd_server_port
      $(METADATA_GRPC_SERVICE_PORT)
      --runtime_info_json
      $(KFP_V2_RUNTIME_INFO)
      --container_image
      $(KFP_V2_IMAGE)
      --task_name
      my-component
      --pipeline_name
      pipeline/my-pipeline
      --run_id
      $(KFP_RUN_ID)
      --run_resource
      workflows.argoproj.io/$(WORKFLOW_ID)
      --namespace
      $(KFP_NAMESPACE)
      --pod_name
      $(KFP_POD_NAME)
      --pod_uid
      $(KFP_POD_UID)
      --pipeline_root
      s3://YOUR_BUCKET/experiments/image/segmentation/baseline
      --enable_caching
      $(ENABLE_CACHING)
      --
      --
    Args:
      sh
      -c
      
      if ! [ -x "$(command -v pip)" ]; then
          python3 -m ensurepip || python3 -m ensurepip --user || apt-get install python3-pip
      fi
      
      PIP_DISABLE_PIP_VERSION_CHECK=1 python3 -m pip install --quiet     --no-warn-script-location 'kfp==1.8.6' && "$0" "$@"
      
      sh
      -ec
      program_path=$(mktemp -d)
      printf "%s" "$0" > "$program_path/ephemeral_component.py"
      python3 -m kfp.v2.components.executor_main                         --component_module_path                         "$program_path/ephemeral_component.py"                         "$@"
      
      
      import kfp
      from kfp.v2 import dsl
      from kfp.v2.dsl import *
      from typing import *
      
      def my_component():
          print("hello")
      
      
      --executor_input
      {{$}}
      --function_to_execute
      my_component
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Sat, 19 Mar 2022 00:46:44 +0000
      Finished:     Sat, 19 Mar 2022 00:47:01 +0000
    Ready:          False
    Restart Count:  0
    Environment Variables from:
      metadata-grpc-configmap  ConfigMap  Optional: true
    Environment:
      AWS_ACCESS_KEY_ID:           <set to the key 'accesskey' in secret 'mlpipeline-minio-artifact'>  Optional: false
      AWS_SECRET_ACCESS_KEY:       <set to the key 'secretkey' in secret 'mlpipeline-minio-artifact'>  Optional: false
      **AWS_REGION:                  us-east-2**
      KFP_POD_NAME:                my-pipeline-khs2z-2408580201 (v1:metadata.name)
      KFP_POD_UID:                  (v1:metadata.uid)
      KFP_NAMESPACE:               dev (v1:metadata.namespace)
      WORKFLOW_ID:                  (v1:metadata.labels['workflows.argoproj.io/workflow'])
      KFP_RUN_ID:                   (v1:metadata.labels['pipeline/runid'])
      ENABLE_CACHING:               (v1:metadata.labels['pipelines.kubeflow.org/enable_caching'])
      KFP_V2_IMAGE:                python:3.7
      KFP_V2_RUNTIME_INFO:         {"inputParameters": {}, "inputArtifacts": {}, "outputParameters": {}, "outputArtifacts": {}}
      ARGO_CONTAINER_NAME:         main
      ARGO_INCLUDE_SCRIPT_OUTPUT:  false
    Mounts:
      /kfp-launcher from kfp-launcher (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4hb8l (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  podmetadata:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.annotations -> annotations
  docker-sock:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/docker.sock
    HostPathType:  Socket
  kfp-launcher:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  mlpipeline-minio-artifact:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  mlpipeline-minio-artifact
    Optional:    false
  kube-api-access-4hb8l:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>
```

**Comment by AlexandreBrown** [comment]
I'm not able to reproduce on a fresh test cluster, I will try to re-install from a fresh cluster for our prod environment and re-open if the issue still occurs.

**Comment by AlexandreBrown** [comment]
I'm not able to reproduce on a fresh test cluster, I will try to re-install from a fresh cluster unless someone has an idea about a fix.  
Will close the issue if re-installing from fresh cluster fixes it.

**Comment by rrrkharse** [comment]
I'll prioritize taking a look and see what is missing that is causing a `301` [response](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/301).

Also I'll take a look at why the S3 URL is for a `us-east-1` object and if that matters.

**Comment by AlexandreBrown** [comment]
Thanks @rrrkharse , I was able to resolve the issue.  
The error was not on AWS side but rather on our side, we solved the issue by following @surajkota 's suggestion and added `s3:GetBucketLocation` to the Kubeflow AWS account permissions (this is only needed if you share a bucket with multiple accounts).  
  
What happened is Kubeflow Pipeline (or something) performs a `GetBucketLocation` at the end of the pipeline when uploading the artifacts. If you don't have `s3:GetBucketLocation`, then it will fall back to the default value `us-east-1` which explains the url and pipeline failure since our bucket was in `us-east-2`.
